Simplify Tensor Parallel implementation with PyTorch TP

kwen2501 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

This PR uses the `torch.distributed.tensor.parallel` subpackage to implement Tensor Parallel for Llama (as an example).

The motivation is multi-fold:

1. to make modeling code simple as single-worker case:
all manual TP implementations under `if self.config.pretraining_tp > 1` can be removed.

3. to make tensor parallelism easily accessible by users:
added a `model.tensor_parallel(device_mesh)` method that allows users to turn a single-proc model into a parallel model. !- Please guide me to a right place to put this function/method if `PreTrainedModel` is not a preferred place. -!

### Note: 
This is just a demo. The removal of `if self.config.pretraining_tp > 1` may break workflows elsewhere, but it is just intended to calculate how much code can be saved, and hopefully it would be possible to direct it to the mechanism in this PR.

### User script:
```
import os
import torch
import torch.distributed as dist

from transformers import AutoTokenizer, AutoModelForCausalLM

model_id = "meta-llama/Meta-Llama-3-8B-Instruct"

def main(rank, world_size) -> None:
    device = torch.device(f"cuda:{rank}")
    dist.init_process_group("nccl", device_id=device)

    model = AutoModelForCausalLM.from_pretrained(
        model_id,
        tp_plan="auto",
    )

    tokenizer = AutoTokenizer.from_pretrained(model_id)
    prompt = "Can I help"

    inputs = tokenizer(prompt, return_tensors="pt").input_ids.to(device)
    outputs = model(inputs)

    next_token_logits = outputs[0][:, -1, :]
    next_token = torch.argmax(next_token_logits, dim=-1)
    response = tokenizer.batch_decode(next_token)
    print(f"{rank=}: {response}")


if __name__ == "__main__":
    main(int(os.environ["RANK"]), int(os.environ["WORLD_SIZE"]))

```

Launch command:
`torchrun --standalone --nproc-per-node 4 tp_hf.py`

Output:
```
Can I help
rank=3: [' you']
rank=1: [' you']
rank=2: [' you']
rank=0: [' you']
```

### Test
```
CUDA_VISIBLE_DEVICES=0,1 pytest -sv tests/tp/test_tp.py
```

### Doc
`perf_infer_gpu_multi.md`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Models:

- text models: @ArthurZucker @younesbelkada @fxmarty 

Library:

- trainer: @muellerzr and @SunMarc

Integrations:

- Big Model Inference: @SunMarc

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)

Cc PyTorch Distributed Team:
@gnadathur @wanchaol @fduwjj @wz337 @wconstab @tianyu-l

kwen2501 committed:
Simplify Tensor Parallel implementation with PyTorch TP

kwen2501 commented:
Cc: [lessw2020](https://github.com/pytorch/torchchat/commits?author=lessw2020) @HamidShojanazeri
Cc: @jerryzh168

ArthurZucker reviewed (commented):
Looks really really nice! I think the only thing I would change is to have the `tp_plan` loadable and potentially defined in the config. Given that it does not hold any weights, it would allow other frameworks to rely on it as well! 🤗

ArthurZucker commented:
Also we would place 
```python
    device = torch.device(f"cuda:{rank}")
    dist.init_process_group("nccl", device_id=device)
    device_mesh = dist.init_device_mesh("cuda", (world_size,))

    with device:
        model = AutoModelForCausalLM.from_pretrained(
            model_id,
        )
```
potentially inside `from_pretrained`, this way we automatically set things for users, with them only having to change the way they call their script (with torchrun or not)

ArthurZucker commented:
WDYT?

ArthurZucker commented:
We will also add this to the `configuration_llama.py`:
```python
DEFAULT_TP_PLAN = {
    "model.layer.*.q_proj":"column_parallel"
}
```
this way we can init lama config with it!

ArthurZucker commented:
cc @SunMarc as per our discussion!

kwen2501 committed:
Move tp_plan to config

kwen2501 commented:
Thanks @ArthurZucker @SunMarc for your review.

Good idea to move `tp_plan` to `configuration_llama.py`. I made that change in the current version. Indeed `modeling_llama.py` becomes cleaner now, and it will not have dependency on `torch.distributed...`. 

Since the top-level model can have different `base_model_prefix` (`"model"`, `"transformer"`, etc), I could not directly put FQNs like `"model.layers.*"` in `configuration_llama.py`. So I prepare a `_base_model_tp_plan` starting with `"layers.*"` facing the base model `LlamaModel`. This works because `model.tensor_parallel()` searches for and applies `_tp_plan` recursively. That way we can serve all models regardless of what prefix name they give to `LlamaModel` or even just a bare-metal `LlamaModel`.

I also made the TP styles "type neutral" in the config as you suggested. Translation to torch TP types is done when applying TP.

kwen2501 commented:
Re moving distributed / device_mesh init code inside `from_pretrained`:

It is a good direction in making the workflow even simpler for users. As in, wouldn't it be nice if the model is distributed after it is created? I think a UI design would be helpful as to how users express the intention of "please create the model in a distributed manner".

For example, would `from_pretrained` accepts a related argument, like:
```
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    device_mesh=device_mesh,
)
```
If that kwarg is given, I think the intention is clear enough, and we can call `tensor_parallel()` for the user. Or even better, initialize partial weights only -- this is supported by the proposed torch TP too, but it will require slightly more work in integration with the weight loading part of `from_pretained`. I can try making a demo in a next PR if you are interested.

Re launcher: I might slightly prefer the "explicit expressioin" to "implicit dependency" on `torchrun`, for the purpose of keeping HF transformers "launcher neutral", as some users may use SLURM, etc. Another reason was that `init_process_group` and `init_device_mesh` can take multiple kwargs (such as timeout, options, etc), which might be a bit big for `from_pretrained` to absorb if it wants to call them inside.

Regardless, I agree that it would be very nice if we can create distributed models out of `from_pretrained" directly, and I am really excite to help that way :)

ArthurZucker commented on src/transformers/pytorch_utils.py:
```diff
@@ -326,3 +331,24 @@ def isin_mps_friendly(elements: torch.Tensor, test_elements: torch.Tensor | int)
         return torch.isin(elements, test_elements)
+
+
+def translate_to_torch_parallel_style(style: str):
+    """
+    In model configurations, we use a neutral type (string) to specify parallel
+    styles, here we translate them into torch.distributed tensor-parallel
+    types.
+    """
+    if not isinstance(style, str):
+        raise ValueError(
+            f"Unsupported parallel style type {type(style)}, expected str"
+        )
+
+    if style == "colwise":
+        return ColwiseParallel()
+    elif style == "rowwise":
+        return RowwiseParallel()
+    elif style == "colwise_rep":
+        return ColwiseParallel(output_layouts=Replicate())
+    else:
+        raise ValueError(f"Unsupported parallel style value: {style}")
```
A mapping from tp style to the correct function might be better. 
We are also thinking of defining a TensorParallelConfig, your feedback is welcome here, as I don't know the variety of classes / args that might be used!

kwen2501 replied:
Thanks for the comment! Indeed a mapping style would look better. 
The only caveat is that the returned value here is an object rather than a constant or class (see the `()` behind `ColwiseParallel`). If we prepare a map, we may be always returning the *same* object -- the `parallelize_module` API may be able to support it I guess, I am just not sure if there is a contract guaranteeing that today.

kmehant replied:
Pitching in :)

We should be able to use the same object since it applies required parallel operation to the module and returns a new copy - https://github.com/pytorch/pytorch/blob/86d4b7d60b264cae5a04a1b20719bcd7a5752a4c/torch/distributed/tensor/parallel/api.py#L95 

Have also tested it empirically while benchmarking (https://github.com/huggingface/transformers/pull/34194)

Thanks!

ArthurZucker replied:
SOunds good!

ArthurZucker reviewed (commented):
Looks a lot better! 
I think we want to have some small tests to make sure we properly serialize the TP-Plan, and also properly dispatch the model (slow tests, we can probably add them ourselves!)
- would need some documentation as well, we can do that in a follow up PR for sure



On of the most important question now is: What's the approach regarding `ForCausalLM`. 

This might be for @SunMarc as it's `transformers` internal: 

`_base_model_tp_plan` will be consumed by the `PreTrainedModel` ➡ this way we can enforce some checks: 
- check that all models that inherit from `PreTrainedModel` have a TP plan 
- check nesting (ForCausalLM has a module that has a TP Plan, if it has other module they should have a TP plan

kmehant commented on src/transformers/models/llama/configuration_llama.py:
```diff
@@ -141,6 +141,16 @@ class LlamaConfig(PretrainedConfig):
     model_type = "llama"
     keys_to_ignore_at_inference = ["past_key_values"]
+    # Default tensor parallel plan for base model `LlamaModel`
+    _base_model_tp_plan = {
```
Do we want to allow this for external use by removing `_` so that we can allow users to define tp plan tweaks from `config.json`? 

Given that, shall we as well allow for providing custom tp plan as input to `LlamaConfig()` that overrides the default?

kwen2501 replied:
Yeah, good idea. We can make this public once we prove things work.

ArthurZucker replied:
Yep, `base_model_tp_plan` should be supported as input to the PreTrainedConfig!

kmehant replied:
So, this variable `base_model_tp_plan` has to be added to PreTrainedConfig https://github.com/huggingface/transformers/blob/a769ed45e17c44fd17b85c025863c4e4f2f73634/src/transformers/configuration_utils.py#L50 with a default value as an empty dict `{}` which I believe is best possible default for any config sub class inheriting.

kwen2501 replied:
Thanks @kmehant @ArthurZucker for the suggestion. I moved `base_model_tp_plan` to `PretrainedConfig` in the latest commit.

kmehant commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1102,6 +1064,7 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
 
 class LlamaForCausalLM(LlamaPreTrainedModel, GenerationMixin):
     _tied_weights_keys = ["lm_head.weight"]
+    _tp_plan = {"lm_head": "colwise_rep"}
```
Trying to think out loud here, 

Do we want keep all the tp_plans as part of the config classes? We pull them here from config as needed, WDYT?

My thinking goes like adding to the LlamaConfig something like

```
llama_for_causal_lm_addon_tp_plan = {"lm_head": "colwise_rep"}
```

So here in the actual model class we pull in like we are pulling in for LlamaModel class (decoder layers)

kwen2501 replied:
That's a good question. The setting in the config class in current PR is a bit conservative so as to align with the config's definition:
```
class LlamaConfig(PretrainedConfig):
    This is the configuration class to store the configuration of a [`LlamaModel`].
```

ArthurZucker replied:
Actually up to debate.
The thing is, `LlamaConfig` is used to init all `LlamaPreTrainedModel`, thus warrants that you can have a need for multiple extra keys (add ons). As such, I think it makes more sense to have them in the class itself. 
But, indeed confining to the config would also be easier. I am just worried about extra add ons that would be "visible" when you have a model that is unrelated (like a SequenceClassification model). 

We can fetch the `_tp_plan` from the class on post init or `super().__init__(config)`

kmehant replied:
> I think it makes more sense to have them in the class itself

Trying to understand, when we confine plans entirely to the model class instead of config classes, how does it allow users to modify tp plans through `config.json`?

An orthogonal thought, do you think we should instead introduce separate config file like we are doing for `FSDP config` where we pass in `fsdp_config` and here it could be `tp_config`?

ArthurZucker replied:
Ah the thing is, we might not want to `serialize` it in the config.json, but they can pass a TP-Plan when initializing as kwargs

ArthurZucker replied:
Yeah we could also think about a separate file, BUT I think going for something close to our `device_map` will be easier. If there is a strong push, then we can have a config as well!

kmehant replied:
I see @ArthurZucker, I thought we wanted to use config.json to both serialise and deserialise for users motivated from your comment - https://github.com/huggingface/transformers/issues/32470#issuecomment-2275768887 :)

kwen2501 commented:
Summarizing my discussion with @kmehant re collaboration plan between this PR and https://github.com/huggingface/transformers/pull/34194:
- Both PRs agree on the interest in having PyTorch TP in transformers :)
- We agree that the integration can hopefully happen at a broad base so that it can scale to a set of models, thus we agree on picking `PretrainedModel` as an integration point.
- [#34194](https://github.com/huggingface/transformers/pull/34194) has changes on the trainer side. And there is a corresponding [PR](https://github.com/huggingface/accelerate/pull/3173) for changes on Accelerate side. So those two PRs focus on workflow-level integration. So hopefully this PR can serve as a base for those two PRs and once landed, unblock the other two PRs.

ArthurZucker commented:
Sounds good, let's focus on this one first then!

kwen2501 committed:
Merge remote-tracking branch 'origin/main' into tp_llama

kwen2501 committed:
Lint

kwen2501 committed:
Format and warning

kwen2501 commented:
Dumb question: is there a way to bypass the `check_repository_consistency` test? Thanks! @ArthurZucker @SunMarc

kwen2501 committed:
Disable copy-from check

kwen2501 commented:
Okay, I just temporarily disabled the "Copied from" comments in some derivative models. Will add them back once we extend TP plans to those models.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_34184). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker reviewed (commented):
Hey! So basically places where you are disabling it , you just need to do something like: 
`# Copied from transformers.models.llama.modeling_llama.LlamaModel with Llama->Olmoe,lm_head->new_name` for example

and before anything run `make fix-copies` which would apply your changes!

kwen2501 commented:
Thanks @ArthurZucker . I tried that method, but it doesn't seem to work. 
Specifically, the test is checking for consistency between `modeling_gemma.py` and `modular_gemma.py`. I edited neither, but the CI complains that they are not consistent. Here is the [link](https://app.circleci.com/pipelines/github/huggingface/transformers/109404/workflows/7fdcf506-4981-470c-8b3c-17e809855f1a/jobs/1453707).

ArthurZucker commented:
Ah that is because you probably modified `modeling` directly, when these two have a `modular` file, which has the source of truth!

kwen2501 committed:
Conditionally get attr from config

kwen2501 committed:
make fix-copies

kwen2501 committed:
Move base_model_tp_plan to PretrainedConfig

kwen2501 committed:
Merge remote-tracking branch 'origin/main' into tp_llama

kwen2501 commented:
@ArthurZucker Fixed code consistency issue via `make fix-copies`. Thanks!
All tests seems passed now (except three awaiting approval). I wonder if you have more comments?
Also cc: @kmehant @SunMarc @muellerzr

kwen2501 committed:
Merge remote-tracking branch 'origin/main' into tp_llama

kmehant commented on src/transformers/modeling_utils.py:
```diff
@@ -5013,6 +5014,41 @@ def _is_quantized_training_enabled(self):
 
         return self.hf_quantizer.is_trainable
 
+    def tensor_parallel(self, device_mesh):
```
@kwen2501 Is there a way to know if tensor_parallel took effect or not? For instances like there was no usable tp plan found in any sub module so no TP took place etc. Returning that indication would help in better downstream usage. 

cc: @ArthurZucker

kwen2501 replied:
I added logging in this method. You can see if by setting `TRANSFORMERS_VERBOSITY=DEBUG`.

kmehant replied:
Logging is great, but need some indication that can be consumed programatically :)

ArthurZucker reviewed (commented):
The only two missing things for me are this:
```
    device = torch.device(f"cuda:{rank}")
    dist.init_process_group("nccl", device_id=device)
    device_mesh = dist.init_device_mesh("cuda", (world_size,))

    with device:
        model = AutoModelForCausalLM.from_pretrained(
            model_id,
        )

```

that should be done here: 
https://github.com/huggingface/transformers/blob/006e86931247d7bacf94ad00ada1fb6baf11e94b/src/transformers/modeling_utils.py#L4075-L4097 

WIth a `tp_plan = "auto"` or `tp_plan =base_model_tp_plan`
And, `PreTrainedModel` that has cls attr _tp_plan default to `config.base_model_tp_plan | cls._tp_plan`

ArthurZucker commented:
I will handle the fix copies afterwards don't worry about it!

kwen2501 commented:
Thanks for your comments @ArthurZucker .
The reason I haven't put `init_process_group` or `init_device_mesh` into `from_pretrained` are two folds:

1. I was not sure if I should make `from_pretrained` a collective API.
Both `init_process_group` and `init_device_mesh` have collectivity requirements. If `from_pretrained` calls them inside, it will have that requirement too. So I was hoping to let `init_device_mesh` undertake that requirement and cut `from_pretrained` loose from it. It also means that when user's program hit a hang, the debug / support load is on PyTorch, not on HF. The user could have also called `init_process_group` before calling `from_pretrained`; if we call it a second time, it may error out.

2. I was not sure if the user would like to shard the model over the entire "world".
In a multi-dimensional scenario, the user may want to TP'lize the model only onto a dimension of the world instead of the entire one (which `init_process_group` creates). This implies that the "slice" must come from the user. Of course, that's not the requirement of the example code, but just want to keep ourselves "future proof." 

That said, I agree that we can probably make the program simpler by moving the `.tensor_parallel()` API into `from_pretrained`. Mixing with 1 and 2, how about the following design:
```
tp_mesh = DeviceMesh(...)
model = AutoModelForCausalLM.from_pretrained(
    model_id, tp_mesh=tp_mesh
)
```
This will achieve both 1 and 2:
- `DeviceMesh(...)` will undertake the collectivity requirement, `from_pretrained` will be free from it.
- User can specify which dimension of the world to apply TP via `tp_mesh`.

kwen2501 commented:
Chatted with @SunMarc offline. Will move the `DeviceMesh` part in.

kwen2501 committed:
Move TP into from_pretrained

kwen2501 committed:
Add device context for load

kwen2501 committed:
Do not serialize

kwen2501 committed:
Move _tp_plan setting to post_init

kwen2501 commented:
Hi @ArthurZucker @SunMarc 
I made the two changes suggested by @ArthurZucker in the comment above.
1. `init_device_mesh` is now moved into `from_pretrained`. When user sets argument `tp_plan="auto"`, the creation will be triggered and `from_pretrained` will return a TP'lized model.
```
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    tp_plan="auto",
)
```
2. `_tp_plan` is now a cls attr of `PreTrainedModel` and I moved `self._tp_plan = config.base_model_tp_plan` from each model's code surface to `PreTrainedModel.post_init()`, so that it is hidden / cleaner.
Would appreciate your review. Thanks!

kwen2501 commented:
And this is benchmarking per @SunMarc suggestion:
![Meta-Llama-3-8B-Instruct, seqlen = 512, python, w_ compile](https://github.com/user-attachments/assets/c8672aca-8b18-441e-abe3-a50fdb33db11)

kwen2501 commented on src/transformers/configuration_utils.py:
```diff
@@ -192,6 +194,7 @@ class PretrainedConfig(PushToHubMixin):
     model_type: str = ""
     is_composition: bool = False
     attribute_map: Dict[str, str] = {}
+    base_model_tp_plan: Optional[Dict[str, Any]] = None
```
Yep done!

kwen2501 commented:
@ArthurZucker @SunMarc Wondering if you could take another look to see if I addressed your comments? Thanks very much!

ArthurZucker commented:
> Hi @ArthurZucker @SunMarc I made the three changes suggested by @ArthurZucker in the comment above.
> 
> 1. `init_device_mesh` is now moved into `from_pretrained`. When user sets argument `tp_plan="auto"`, the creation will be triggered and `from_pretrained` will return a TP'lized model. That is, the following UI is now supported:
> 
> ```
> model = AutoModelForCausalLM.from_pretrained(
>     model_id,
>     tp_plan="auto",
> )
> ```
> 
> 2. `_tp_plan` is now a cls attr of `PreTrainedModel` and I moved `self._tp_plan = config.base_model_tp_plan` from each model's code surface to `PreTrainedModel.post_init()`, so that it is hidden / cleaner.
> 3. Removed `base_model_tp_plan` from `to_dict` and `to_diff_dict` so that it won't be serialized / saved atm.
> 
> Would appreciate your review. Thanks!

Sounds absolutely great! :hugs:

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -1442,6 +1449,9 @@ def post_init(self):
         self.init_weights()
         self._backward_compatibility_gradient_checkpointing()
+        # If current model is a base model, attach `base_model_tp_plan` from config
+        if self.base_model is self:
```
```suggestion
        if self._tp_plan is None:
```

ArthurZucker replied:
feels simpler

ArthurZucker replied:
Yeah, but `self._tp_plan` should be None only for the base model no? 
We could maybe add something like `if not self.base_model is self and self._tp_plan is None and self.supports_tp_plan` raise an error in the futur, to force people to add the TP plan.

kwen2501 replied:
> but self._tp_plan should be None only for the base model no?

That's not entirely the case. For example, `LlamaForSequenceClassification` and `LlamaForQuestionAnswering` have `_tp_plan=None` (at its top level), while `LlamaForCausalLM` has a `_tp_plan = {"lm_head": ...}`.

ArthurZucker replied:
Yep, indeed.  I thiught we should enforce TP plan definition for all classes to avoid user errors but its fine like this!

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -4324,6 +4357,14 @@ def from_pretrained(
             return model, loading_info
 
+        if tp_plan is not None:
+            assert tp_device is not None, "tp_device not set!"
```
let's raise n error instead

kwen2501 replied:
Oh, here the assert is more for catching internal error than user error.
It is like "I swear this cannot happen” but it seems to happen anyway.
If someone runs python with `-O` flag, this assert can be optimized away :)

ArthurZucker replied:
Okay!

ArthurZucker reviewed (approved):
I love it! TBH We can almost merge as is, the main issue / next PR is adding tests. 
We need to make sure there are no weird interactions! 🤗 
Let's add the graph to the doc somewhere as well !

kmehant commented:
@kwen2501 @ArthurZucker 
Do we want to address this minor support - https://github.com/huggingface/transformers/pull/34184#discussion_r1827185086 in this PR? This will unblock accelerate PR.

kwen2501 commented:
@kmehant Thanks for the suggestion. Understood. Agree with the need for a programatic approach.
For the `.tensor_parallel` method itself, it may not be very easy to return a True/False because `nn.Module.apply` does not return anything than the module itself.
Workflow-wise, I wonder if it would be clearer / safer to do:
```
if model.has_tp_plan:
    model.tensor_parallel()
```
If that addresses the need, I am happy to add a `has_tp_plan` property in a follow-up PR.

kwen2501 commented:
> I love it! TBH We can almost merge as is, the main issue / next PR is adding tests. We need to make sure there are no weird interactions! 🤗 Let's add the graph to the doc somewhere as well !

Good idea. Just looked [here](https://huggingface.co/docs/transformers/testing#distributed-training), seems easy to follow the example.

kmehant commented:
https://github.com/huggingface/transformers/pull/34184#issuecomment-2475352157

@kwen2501 Absolutely, this will be helpful, Thanks

ArthurZucker commented:
Sounds good!

kwen2501 committed:
Add has_tp_plan

kwen2501 committed:
Add test_tp

kwen2501 committed:
Add 'Multi-gpu inference' doc

kwen2501 committed:
Merge remote-tracking branch 'origin/main' into tp_llama

kwen2501 committed:
Add backward support for device type identification

kwen2501 commented:
@ArthurZucker Thanks again for the suggestions.
1. A test has been added at `tests/tp/test_tp.py`, run command:
```
CUDA_VISIBLE_DEVICES=0,1 pytest -sv tests/tp/test_tp.py
```
2. A new doc has been added: `perf_infer_gpu_multi.md`.
A use example and the benchmark figure have been added to the doc.
3. @kmehant The `.has_tp_plan` property has been added.

kwen2501 commented:
@ArthurZucker Can you please help merge this image upload for doc? Thanks!
https://huggingface.co/datasets/huggingface/documentation-images/discussions/392

ArthurZucker commented:
Merged! let me review one last time 🤗

ArthurZucker commented on docs/source/en/perf_infer_gpu_multi.md:
```diff
@@ -0,0 +1,68 @@
+<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

ArthurZucker commented on docs/source/en/performance.md:
```diff
@@ -53,7 +53,7 @@ sections we go through the steps to run inference on CPU and single/multi-GPU se
 * [Inference on a single CPU](perf_infer_cpu)
 * [Inference on a single GPU](perf_infer_gpu_one)
-* [Multi-GPU inference](perf_infer_gpu_one)
+* [Multi-GPU inference](perf_infer_gpu_multi)
```
we should not delete! Both can cohexist!

ArthurZucker replied:
Alright thanks!

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -4095,6 +4111,26 @@ def from_pretrained(
+                    device_type = "cuda"
+                else:
+                    raise RuntimeError(
+                        "Device type unknown. Please run model.tensor_parallel with an explicit DeviceMesh."
```
I am not sure I see how this is actionnable? If you pass `tp_plan="auto"` but have a lower version of torch and no cuda how do you bypass this?

ArthurZucker replied:
let's make it more actionnable please!

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -5026,6 +5078,54 @@ def _is_quantized_training_enabled(self):
         return self.hf_quantizer.is_trainable
 
+    @property
+    def has_tp_plan(self):
```
```suggestion
    def supports_tp_plan(self):
```

ArthurZucker commented on tests/tp/test_tp.py:
```diff
@@ -0,0 +1,91 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

ArthurZucker commented on tests/tp/test_tp.py:
LGTM, would be nice to have a generic way for all models that have `support_tp_plan=True` instead of just llama, setting everyone up for success in adding support for other models!

kwen2501 replied:
Yeah, both me and @kmehant are interested in extending this mechanism to other models. We can think of an way to "copy" the tp_plan to other models with similar architectures (just like how HF copies modeling code between similar models). Maybe covering all `AutoModelForCausalLM` would be a good next step. Follow-up PR will come soon :)

ArthurZucker reviewed (approved):
Great work! 🔥 some small things left to do but super close to merge!

kwen2501 committed:
Auto-detect accelerator

kwen2501 committed:
supports_tp_plan

kwen2501 committed:
copyright year

kwen2501 reviewed (commented):
Thanks @ArthurZucker for the close look! I fixed the places you pointed out.

kwen2501 committed:
Merge remote-tracking branch 'origin/main' into tp_llama

kwen2501 committed:
Merge branch 'main' into tp_llama

kwen2501 committed:
Fix copy

kwen2501 commented:
@ArthurZucker CI is all green now. (Previous failure seems to come from infra instability). Is there anything else I should add before merge? Thanks!

ArthurZucker commented:
Nope merging! 🤗

ArthurZucker commented:
I wanted to check if `torch._C.get_device_accelerator` had been available for a while or not!

ArthurZucker commented:
SUper kudos for the PR! 🚀

loadams commented:
Hi @ArthurZucker and @kwen2501 - it seems this PR now requires that users have torch 2.5+, is that corret?  When trying to run the DeepSpeed unit tests with the latest transformers and Torch 2.3 I'm getting errors around not being able to import from `torch.distributed.tensor`:

```
from torch.distributed.tensor import Replicate
```

kmehant commented:
@ArthurZucker @kwen2501 Planning for a HF blog? I am in!

kwen2501 commented:
@loadams Thanks for reporting it and sorry about the break. 
I think that's because `Replicate` was previously in `torch.distributed._tensor`.
Let me open a PR to build backward compat support.
Cc: @wz337

ArthurZucker commented:
Sure! Won't have bandwidth to start a blog but would love to see one!
